### ISSUE TITLE
'evhttp_set_max_body_size()' and reset connections

### ISSUE THREAD

**Issue by nntrab** [libevent/libevent#321]
If a maximum body size is specified via a call to 'evhttp_set_max_body_size()' and a POST request is made that exceeds that size, the connection is reset, rather than an error being displayed.

This was tested on the master branch as of 2016-01-31 on CentOS 7. The file 'sample/https-server.c' had one line added at line 368:

`      evhttp_set_max_body_size(http,` 64\*1024\*1024);`

The code was then compiled and run. The server' root directory had a simple HTML file containing a basic form consisting of a file input and a submit input ([postTest.html](https://github.com/libevent/libevent/files/112795/postTest.zip)). If a file smaller then 64MB was sent then the file was accepted and a blank screen was returned, as per the sample code. Any file larger than 64MB resulted in the Firefox browser displaying a connection reset message. Attempting the same via Chrome and IE on a Windows PC displayed similar behaviour.

In tracing through the code, the library calls 'evhttp_send_error()' with an error code of 400. Strictly speaking I think that should be 413 but regardless the client doesn't receive any error at all, at least according to Firefox's network dialog.

While tracing through the code I also noticed that the 'content-length' check at line 2132 in http.c doesn't ever appear to be invoked, given that none of the browsers tested sent an 'Expect' header. This would imply that the maximum allowed body would have to be received before the client would be sent any error, even if the 'content-length' clearly indicates more data than is allowed is to be sent.

**Comment by azat** [comment]
Hi @nntrab,

Indeed, server need to respond with 413 instead of 400 in this case.
But server don't just reset the connection on error, it sends 400 error, though I didn't test http server with firefox as a client:
```
  connect(10, {sa_family=AF_INET, sin_port=htons(43877), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress) <-- client
  ...
  accept4(9, {sa_family=AF_INET, sin_port=htons(43908), sin_addr=inet_addr("127.0.0.1")}, [16], SOCK_CLOEXEC|SOCK_NONBLOCK) = 11 <-- server
  writev(10, [{"POST / HTTP/1.1\r\nHost: somehost\r\nContent-Length: 8191\r\n\r\n", 57}, {"a {REPEATED 8191 TIMES}", 8191}], 2) = 8248
  ...
  readv(11, [{"POST / HTTP/1.1\r\nHost: somehost\r\nContent-Length: 8191\r\n\r\na {REPEATED 8191 TIMES}", 4096}], 1) = 4096
  ...
  writev(11, [{"HTTP/1.1 400 Bad Request\r\nContent-Type: text/html\r\nConnection: close\r\nDate: Fri, 05 Feb 2016 07:08:42 GMT\r\nContent-Length: 94\r\n\r\n", 129}, {"<HTML><HEAD>\n<TITLE>400 Bad Request</TITLE>\n</HEAD><BODY>\n<H1>Bad Request</H1>\n</BODY></HTML>\n", 94}], 2) = 223
  ...
  readv(10, [{"HTTP/1.1 400 Bad Request\r\nContent-Type: text/html\r\nConnection: close\r\nDate: Fri, 05 Feb 2016 07:08:42 GMT\r\nContent-Length: 94\r\n\r\n<HTML><HEAD>\n<TITLE>400 Bad Request</TITLE>\n</HEAD><BODY>\n<H1>Bad Request</H1>\n</BODY></HTML>\n", 223}], 1) = 223
```

As for Expect header we have 413 error for it, plus we send body along with headers:
```
  connect(10, {sa_family=AF_INET, sin_port=htons(43877), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress) <-- client
  accept4(9, {sa_family=AF_INET, sin_port=htons(51064), sin_addr=inet_addr("127.0.0.1")}, [16], SOCK_CLOEXEC|SOCK_NONBLOCK) = 11 <-- server
  ...
  writev(10, [{"POST / HTTP/1.1\r\nHost: somehost\r\nExpect: 100-continue\r\nContent-Length: 8191\r\n\r\n", 79}, {"a {REPEATED 8191 TIMES}", 8191}], 2) = 8270
  ...
  readv(11, [{"POST / HTTP/1.1\r\nHost: somehost\r\nExpect: 100-continue\r\nContent-Length: 8191\r\n\r\na {REPEATED 8191 TIMES}", 4096}], 1) = 4096
  ...
  writev(11, [{"HTTP/1.1 413 Request Entity Too Large\r\nContent-Type: text/html\r\nConnection: close\r\nDate: Fri, 05 Feb 2016 07:08:42 GMT\r\nContent-Length: 120\r\n\r\n", 143}, {"<HTML><HEAD>\n<TITLE>413 Request Entity Too Large</TITLE>\n</HEAD><BODY>\n<H1>Request Entity Too Large</H1>\n</BODY></HTML>\n", 120}], 2) = 263
  ...
  readv(10, [{"HTTP/1.1 413 Request Entity Too Large\r\nContent-Type: text/html\r\nConnection: close\r\nDate: Fri, 05 Feb 2016 07:08:42 GMT\r\nContent-Length: 120\r\n\r\n<HTML><HEAD>\n<TITLE>413 Request Entity Too Large</TITLE>\n</HEAD><BODY>\n<H1>Request Entity Too Large</H1>\n</BODY></HTML>\n", 263}], 1) = 263
```

**Comment by nntrab** [comment]
I changed the max body size down to the 8kB that you used for your test. If I sent a file that is slightly larger than the limit, I did get a 'Bad Request' page. However, if I sent a much bigger file I got the 'Connection Reset' message. Having done some experimenting, the threshold varies wildly and isn't even consistent from the same machine and browser. The lowest figure I found was about 32kB for Firefox running on the same CentOS 7 VM as the server and the highest was about 2MB from Firefox running on a Windows 7 PC. 

My comments about the 'Expect' header were not that you have an error for it, rather that the test won't be invoked for connections coming from at least 3 of the more modern browsers and therefore the check to see if the 'Content-Length' exceeds the max body size won't be performed at the earliest opportunity. I suppose the question really is why is the content length test only being done if an 'Expect' header is sent and the request version is 1.1?

**Comment by azat** [comment]
On Fri, Feb 05, 2016 at 04:41:01AM -0800, nntrab wrote:
> I changed the max body size down to the 8kB that you used for your test. If I sent a file that is slightly larger than the limit, I did get a 'Bad Request' page. However, if I sent a much bigger file I got the 'Connection Reset' message. Having done some experimenting, the threshold varies wildly and isn't even consistent from the same machine and browser. The lowest figure I found was about 32kB for Firefox running on the same CentOS 7 VM as the server and the highest was about 2MB from Firefox running on a Windows 7 PC. 

There are some issues with larger payloads, but mostly on the client
side, anyway can you please test next patch (also it will be perfect to
have results for curl and various browsers (including chrome))?
519786c18caf77912592aeff122433d3294dd28b

If this will not help, then I guess we will need to implement something
like 'lingering close'.

> My comments about the 'Expect' header were not that you have an error for it, rather that the test won't be invoked for connections coming from at least 3 of the more modern browsers and therefore the check to see if the 'Content-Length' exceeds the max body size won't be performed at the earliest opportunity. I suppose the question really is why is the content length test only being done if an 'Expect' header is sent and the request version is 1.1?

Indeed, that is what I'm talking about about, but this is another story
(and fixing this is very simple), let's fix the connection-reset issues
first.

**Comment by azat** [comment]
Also you can get changes (including mentioned patch and patch for unit test) with the following command:
``` git clone --branch more-graceful-http-v4 git://github.com/azat/libevent.git ```

**Comment by nntrab** [comment]
## Introduction
The patched source was downloaded to a CentOS 7 VM running on a Windows Vista PC. The line 'evhttp_set_max_body_size(http, 8*1024)' was added to 'http-server.c' and the software was compiled. The http-server executable with was then run on the VM and various combinations of client and OS were tested using a simple HTML file for submitting a file.

In all tests the thresholds at which the server errors were no longer displayed are not absolute. Repeated attempts would sometimes display the server's errors and sometimes the browser's connection errors. The mentioned threshold were when connection errors were nearly always displayed. 

## Tests
**Firefox on the CentOS VM the http-server process is running on**
'The connection has been reset' errors appeared as soon as the uploaded file reached 32kB. Below that size the server's 'Bad Request' errors were displayed, rather than 'Entity Too Large' errors.

**Firefox on the Windows Vista PC running the CentOS VM**
'The connection has been reset' errors appeared as soon as the uploaded files reached 128kB. Below that size the server's 'Bad Request' errors were displayed.

**Chrome on the Windows Vista PC running the CentOS VM**
'ERR_CONNECTION_RESET' errors appeared as soon as the uploaded files reached 64kB. Below that size the server's 'Bad Request' errors were displayed.

**MSIE 9 on the Windows Vista PC running the CentOS VM**
"IE cannot display the webpage' errors appeared as soon as the uploaded files reached 128kB. Below that size the server's 'Bad Request' errors were displayed.

**Firefox on a Windows 7 PC on the same LAN**
'The connection has been reset' errors appeared as soon as the uploaded files reached 384kB. Below that size the server's 'Bad Request' errors were displayed. Interestingly this figure was much lower than the figure I saw the last time I ran this test.

**Chrome on a Windows 7 PC on the same LAN**
'ERR_CONNECTION_ABORTED' errors appeared as soon as the uploaded files reached 384kB. Below that size the server's 'Bad Request' errors were displayed. Interestingly the error was not the same one as Chrome running on the Windows Vista PC.

**MSIE 11 on a Windows 7 PC on the same LAN**
'Bad Request' errors are always displayed when the uploaded files exceeded 8kB. The error appeared almost instantly regardless of file size.

**Curl 7.47.1 on a Windows 7 PC on the same LAN**
The server's 'Request Entity Too Large' errors appeared as soon as the uploaded files exceeded 8kB. The error appeared almost instantly regardless of file size.

## Summary
Only Curl displayed the correct error. Of the normal browsers, only MSIE 11 always displayed the server errors. Is this 'bug' due to most browsers only listening for a response once the request has been fully sent? If MSIE 11 were the first to start listening before starting transmission of the request, it would explain why it (mostly) behaved as expected. Does Libevent need to read but discard data until the request has been fully transmitted by the client and only then send the error? Is this what you were referring to when you mentioned 'lingering close'

**Comment by azat** [comment]
Many thanks for testing and results!

Yes by lingering close I mean read all response always (but I'm not sure
100% in this, maybe we need to read only up to the end of the chunk for
example), nginx also have something like this, but this is not what I
mean in this case (since nginx always handle requests greater
`client_max_body_size` regardless `lingering_close` setting, and it's
meaning of `lingering_close` is different).

I do some testing by my self, and google-chrome on linux works fine,
while google-chrome on win32 doesn't work correctly (by correctly I mean
returning bad-request error, even though this is not correct http code
for this case, but this is another story and must be fixed separately,
plus after this patches will be completed the only thing we need to do
for returning entity-too-large instead is changing http code).

So I wrote a patch that will read all the client body even if it exceed
`max_body_size` -- f83adcb (attached to this issue).

Can you test it?

**Comment by azat** [comment]
@nntrab okay I have full patch set that have this stuff under flags and return 413 always (IOW even if you don't have "Expect: 100-Continue", can you please test it:
```
git clone --branch more-graceful-http-v8 https://github.com/azat/libevent
```

And here is piece of code that you need for *server*:
```c
evhttp_set_flags(http, EVHTTP_SERVER_LINGERING_CLOSE);
```

**Comment by nntrab** [comment]
I've tested again on all the previous combinations of browser, OS and machine. All are now correctly returning an Request Entity Too Large error regardless of the file size. This latest patch appears to have fixed the problem. Thank you :)

**Comment by azat** [comment]
@nntrab merged!
